builder/bundle_control_test.go

```diff
@@ -0,0 +1,432 @@
+package builder
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"testing"
+
+	"github.com/clearlinux/mixer-tools/config"
+)
+
+func TestGetUpstreamBundlesVerDir(t *testing.T) {
+	testCases := []struct {
+		ver string
+		exp string
+	}{
+		{"0", "clr-bundles-0"},
+		{"000001", "clr-bundles-000001"},
+		{"non-numeric", "clr-bundles-non-numeric"},
+		{"", "clr-bundles-"},
+		{"25660", "clr-bundles-25660"},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.ver, func(t *testing.T) {
+			actual := getUpstreamBundlesVerDir(tc.ver)
+			if actual != tc.exp {
+				t.Errorf("expected %s on input %s but got %s", tc.exp, tc.ver, actual)
+			}
+		})
+	}
+}
+
+func TestGetUpstreamBundlesPath(t *testing.T) {
+	testCases := []struct {
+		ver string
+		exp string
+	}{
+		{"0", "test/upstream-bundles/clr-bundles-0/bundles"},
+		{"000001", "test/upstream-bundles/clr-bundles-000001/bundles"},
+		{"non-numeric", "test/upstream-bundles/clr-bundles-non-numeric/bundles"},
+		{"", "test/upstream-bundles/clr-bundles-/bundles"},
+		{"25660", "test/upstream-bundles/clr-bundles-25660/bundles"},
+	}
+	b := New()
+	b.Config = config.MixConfig{}
+	b.Config.LoadDefaultsForPath(false, "test")
+	b.Config.Builder.VersionPath = "test"
+	for _, tc := range testCases {
+		t.Run(tc.ver, func(t *testing.T) {
+			actual := b.getUpstreamBundlesPath(tc.ver)
+			if actual != tc.exp {
+				t.Errorf("expected %s on input %s but got %s", tc.exp, tc.ver, actual)
+			}
+		})
+	}
+}
+
+func TestGetLocalPackagesPath(t *testing.T) {
+	b := New()
+	// internally this function requires some sub-objects of builder to
+	// be set (non-nil)
+	b.Config = config.MixConfig{}
+	b.Config.LoadDefaultsForPath(false, "test")
+	b.Config.Builder.VersionPath = "test"
+	b.LocalPackagesFile = "local-packages"
+	actual := b.getLocalPackagesPath()
+	if actual != "test/local-packages" {
+		t.Errorf("expected test/local-packages but got %s", actual)
+	}
+}
+
+func TestGetUpstreamPackagesPath(t *testing.T) {
+	testCases := []struct {
+		ver string
+		exp string
+	}{
+		{"0", "upstream-bundles/clr-bundles-0/packages"},
+		{"000001", "upstream-bundles/clr-bundles-000001/packages"},
+		{"non-numeric", "upstream-bundles/clr-bundles-non-numeric/packages"},
+		{"", "upstream-bundles/clr-bundles-/packages"},
+		{"25660", "upstream-bundles/clr-bundles-25660/packages"},
+	}
+
+	b := New()
+	for _, tc := range testCases {
+		t.Run(tc.ver, func(t *testing.T) {
+			b.UpstreamVer = tc.ver
+			actual := b.getUpstreamPackagesPath()
+			if actual != tc.exp {
+				t.Errorf("expected %s but got %s", tc.exp, actual)
```

Comment:
i.e I'd say doing for this what you did on line #69 is totally enough IMHO.

Reply:
I think if either test should be modified it should be the one on L69 to provide more cases. I don't see the value in *taking away* test cases that have already been written.

Reply:
_ok then_
I just want to reiterate having _useful_ test cases instead of just extra lines of code going forward.

Reply:
Well I'd argue that all of those test-cases are useful. This is the difference between *white-box testing* and *black-box testing*. When black-box testing we can't assume we know the internals of the function at all *or* that the internals will remain the same. Those test cases are essentially black-box tests to make sure we are always testing the expected result even if the internals of the function changes. Black box tests are not "extra lines of code" and are valuable even if you personally know the internal workings of the functions under test.